### PR TITLE
maxbin2 2.2.5 (new formula)

### PR DESCRIPTION
**Pull request by suujia**
- [x] Have you followed the [guidelines for contributing](https://github.com/brewsci/homebrew-bio/blob/master/CONTRIBUTING.md)?
- [x] Have you checked that there aren't other open [pull requests](https://github.com/brewsci/homebrew-bio/pulls) for the same formula update/change?
- [ ] Have you built your formula locally with `brew install --build-from-source FORMULA`, where `FORMULA` is the name of the formula you're submitting?
- [x] Does your build pass `brew audit --strict FORMULA` (after doing `brew install FORMULA`)?

-----

### CODE REVIEW

**Comment by suujia** [comment]
Still testing! 
From read me: 
==== 
```
Enter src directory under MaxBin and "make" it.

=== Prerequisite ===
1. g++ (c++11 or higher), make
2. Perl5 with CPAN, LWP::Simple, and FindBin modules
   (./autobuild_auxiliary will use CPAN to check and install the other two Perl modules)
3. (for automatic downloading and installing auxiliary software) curl, unzip, which

   - The manual way. Auxiliary packages include Bowtie2, FragGeneScan, IDBA-UD, and Hmmer3. Please follow the instructions in the software packages to compile and install them.

	* Bowtie2 (tested version: bowtie2-2.2.3)
	  http://bowtie-bio.sourceforge.net/bowtie2/index.shtml
	* FragGeneScan (tested version: FragGeneScan 1.30)
	  http://omics.informatics.indiana.edu/FragGeneScan/
	* Hmmer3 (tested version: HMMER 3.1b1 64 bit)
	  http://hmmer.janelia.org/
	* IDBA-UD (tested version: IDBA-UD 1.1.1)
	  http://i.cs.hku.hk/~alse/hkubrg/projects/idba_ud/
```

File: Formula/maxbin2.rb [hunk 1]
```diff
@@ -0,0 +1,39 @@
+class Maxbin2 < Formula
+    # cite Wu_2016: "https://doi.org/10.1093/bioinformatics/btv638"
+    desc "an automated binning algorithm to recover genomes from multiple metagenomic datasets"
+    homepage "https://bioconda.github.io/recipes/maxbin2/README.html"
+    url "https://sourceforge.net/projects/maxbin2/files/MaxBin-2.2.5.tar.gz/download"
```
Comment (sjackman) [inline Formula/maxbin2.rb]: Try…
```ruby
url "https://downloads.sourceforge.net/project/maxbin2/MaxBin-2.2.5.tar.gz"
```

File: Formula/maxbin2.rb [hunk 2]
```diff
@@ -0,0 +1,39 @@
...
+    depends_on "cpanminus" => :build
+    depends_on "bowtie2"
+    depends_on "hmmer"
+    depends_on "idba"
+    depends_on "perl" unless OS.mac?
+  
+    resource "fraggenescan" do 
+        url "https://sourceforge.net/projects/fraggenescan/files/FragGeneScan1.31.tar.gz/download"
+        sha256 "cd3212d0f148218eb3b17d24fcd1fc897fb9fee9b2c902682edde29f895f426c"
+    end
```
Comment (sjackman) [inline Formula/maxbin2.rb]: Could `fraggenescan` be a separate formula rather than a resource?

File: Formula/maxbin2.rb [hunk 3]
```diff
@@ -0,0 +1,39 @@
...
+
+    def install
+      ENV.prepend "PERL5LIB", Formula["bioperl"].libexec/"lib/perl5"
```
Comment (sjackman) [inline Formula/maxbin2.rb]: Use `ENV.prepend.path` to ensure that the separator is `:` rather than a space

File: Formula/maxbin2.rb [hunk 4]
```diff
@@ -0,0 +1,39 @@
...
+      system "cpanm", "-l", prefix/"perl5", "LWP::Simple"
+  
+      resource("fraggenescan").stage do
+        system "make", "clean"
+        system "make", "fgs"
+      end
+
+      cd "src" do
+          system "make"
+      end
```
Comment (sjackman) [inline Formula/maxbin2.rb]: After running `make` you'll need to install the executables that it built. If there's no `make install` target, then use…
```ruby
bin.install "maxbin2", "any-other-executables-go-here"
```

**Comment by sjackman** [comment]
Check the output of `brew style maxbin2` and `brew audit --strict --new-formula maxbin2`

**Comment by tseemann** [comment]
```
==> brew readall --aliases brewsci/bio
Error: No available formula with the name "fraggenescan" 
Exited with code 1
```

**Comment by suujia** [comment]
Hmm. not really sure what to do about this error. 
```
==> brew install --verbose --build-bottle brewsci/bio/maxbin2
==> Installing maxbin2 from brewsci/bio
==> Downloading https://downloads.sourceforge.net/project/maxbin2/MaxBin-2.2.5.tar.gz
Already downloaded: /home/linuxbrew/.cache/Homebrew/maxbin2--2.2.5.tar.gz
==> Verifying maxbin2--2.2.5.tar.gz checksum
gunzip -N -- /tmp/d20180830-6755-nyyvzs/MaxBin-2.2.5.tar.gz
tar xf /tmp/d20180830-6755-nyyvzs/MaxBin-2.2.5.tar -C /tmp/d20180830-6755-11lpicc
cp -pR /tmp/d20180830-6755-11lpicc/MaxBin-2.2.5/. /tmp/maxbin2-20180830-6755-1lxk5r6/MaxBin-2.2.5
Error: wrong number of arguments (given 0, expected 2..3)
```

File: Formula/maxbin2.rb [hunk 5]
```diff
@@ -0,0 +1,30 @@
...
+  depends_on "cmake" => :build
+  depends_on "cpanminus" => :build
+  depends_on "bowtie2"
+  depends_on "fraggenescan"
+  depends_on "hmmer"
+  depends_on "idba"
+  depends_on "perl" unless OS.mac?
+
+  def install
+    ENV.prepend.path "PERL5LIB", Formula["bioperl"].libexec/"lib/perl5"
```
Comment (sjackman) [inline Formula/maxbin2.rb]: There's a typo here. `prepend.path` should be `prepend_path`

File: Formula/maxbin2.rb [hunk 6]
```diff
@@ -0,0 +1,30 @@
...
+    ENV.prepend_create_path "PERL5LIB", prefix/"perl5/lib/perl5"
+    system "cpanm", "-l", prefix/"perl5", "LWP::Simple"
+    cd "src" do
+      bin.install Dir["*"]
+    end
+    bin.install_symlink Dir["#{prefix}/run_MaxBin.pl"]
```
Comment (sjackman) [inline Formula/maxbin2.rb]: `Dir` is only needed when using a wildcard like `*`
```ruby
    bin.install_symlink prefix/"run_MaxBin.pl"
```

**Comment by suujia** [comment]
Not sure how to fix this! 
```
Program not built. Please run "make" under src directory to build MaxBin program.
```
I've tried it with  bin.install Dir["*"] following the system"make" as well. This is the makefile in src: 
```
CC:=g++
CFLAGS:=-c -Wall -g -std=c++11
LDFLAGS:=-pthread -std=c++11
INC:=-I ./
SOURCES:=AbundanceLoader.cpp EManager.cpp EucDist.cpp fastaReader.cpp \
	KendallDist.cpp kmerMap.cpp logger.cpp main.cpp ManhattanDist.cpp \
	MinManhattanDist.cpp Profiler.cpp quickSort.cpp SpearmanDist.cpp \
	AbstractDist.cpp NormalDistribution.cpp ThreadPool.cpp
OBJECTS:=$(SOURCES:.cpp=.o)
EXECUTABLE:=MaxBin
all: $(EXECUTABLE)
$(EXECUTABLE): $(OBJECTS)
	$(CC) $(LDFLAGS) -o $(EXECUTABLE) $(OBJECTS) $(INC) $(LDFLAGS)
$(OBJECTS): %.o: %.cpp
	$(CC) $(CFLAGS) $< -o $@ $(INC)
clean:
	rm $(EXECUTABLE) $(OBJECTS)
```
and the readme just says
```
Enter src directory under MaxBin and "make" it.
```


### Other changes I've made: 
added LWP::Simple based on this issue: https://github.com/brewsci/homebrew-bio/pull/267#issuecomment-388999848
added IO::Socket::SSL based on this issue: https://github.com/brewsci/homebrew-bio/pull/267#issuecomment-397708020

**Comment by suujia** [comment]
Build passed on linux and travis, fails on mac for following error: 
```
  Test::Unit::AssertionFailedError: </Usage/> expected to be =~
<"sh: line 1: 25890 Abort trap: 6           idba_ud > /dev/null 2> tmp_1536130889\nCannot run IDBA-UD. Please indicate the file directory in 'setting' file.\n">.
```

In specifications it says: 
prerequisites:  g++ (c++11 or higher)  

Should we specify anything relating to that?

**Comment by sjackman** [comment]
```
❯❯❯ idba_ud
dyld: Library not loaded: /Users/sjackman/.homebrew/opt/gcc/lib/gcc/7/libgomp.1.dylib
  Referenced from: /Users/sjackman/.homebrew/bin/idba_ud
  Reason: image not found
```
The `idba` bottle needs to be rebuilt on macOS for `gcc`.
I've opened PR https://github.com/brewsci/homebrew-bio/pull/421

**Comment by sjackman** [comment]
I've merged PR #421 to bump `idba`, and I've rebased this PR onto master.

**Comment by sjackman** [comment]
Merged. Thanks again, Susie! 🏆

**Comment by sjackman** [comment]
I've added `brew update` to `orca-5` and triggered a rebuild on Docker Hub.

**Pull request closed without merging**